### DISCUSSION TITLE
Double canvases

### DISCUSSION THREAD

**Discussion by bryevdv** [bokeh/bokeh#12140] [Internals and design (Q&A)] [answered]
Just noticed in both 2.4 and 3.0 docs that gallery examples seem to have a second blank canvas under the canvas that is drawn on:

<img width="1801" alt="Screen Shot 2022-05-22 at 15 40 40" src="https://user-images.githubusercontent.com/1078448/169719109-5fafeeb4-85e5-4611-bc8b-b366bf570645.png">

These are "autoload" embeds but also seems the case in bokeh server apps and `output_file` output. I am probably just forgetting some change in the past, but want to make sure. Is this intentional? If so, what is the rationale? cc @mattpap (or @ianthomas23 if you have some idea)

**Comment by ianthomas23** [comment, marked as answer]
A quick look through the code shows that there are indeed 2 canvases:

https://github.com/bokeh/bokeh/blob/530c735f75f065b33acced0096f58e265aae4a65/bokehjs/src/lib/models/canvas/canvas.ts#L97-L98

A relevant PR is #9948.

**Comment by mattpap** [comment]
Historically we used single canvas and HTML based annotations for overlays, but that was cumbersome at best (e.g. keeping canvas and HTML versions of those annotations in line with each other, if possible at all). Currently there are two canvases, the primary one for everything except overlays and one for overlays. This way, rapidly changing overlays (e.g. crosshair tool's spans) don't require expensive repaint of glyphs, etc. We could keep offscreen canvases and then paint those on a single canvas, but that would result in flickering if simple overpaint algorithm is used. Implementing anything more sophisticated seems to be more trouble than it's worth, given how well the current setup works. Also it's worth to note that `ctx.drawImage()` is actually pretty expensive. In PR #12083 the existing approach will be generalized further, to allow arbitrary number of user-defined layers.
